### ISSUE TITLE
Get all shell scripts passing shellcheck

### ISSUE THREAD

**Issue by Ericson2314** [NixOS/nix#10795]
Since #10787, we are now running shellcheck on scripts.

Just like with C/C++ formatting, we are excluding almost all existing files by default. However, the reason for doing that is different than for formatting --- fixing shellcheck lint errors does *not* have the downside of causing merge conflicts churn, but does require manual edits. Thus we're skipping existing files *not* because it would be potentially bad to get them passing, but because it is more work than is appropriate for a single PR.

Scripts can one-by-one be made passing, which is a great issue to share between new contributors!

**Comment by SkamDart** [comment]
Is there a way to run individual tests?


```bash
#!/usr/bin/env bash

# this file was made by manually editing the list of the shcellcheck exluded files.
#
# Use some vim/emacs magic from the output of
#
# $ cat $_NIX_PRE_COMMIT_HOOKS_CONFIG | yq -y .repos[0].hooks[1].exclude
#
# e.g.
# tests/functional/gc.sh
# tests/functional/add.sh
# tests/functional/fmt.sh
FILES=$(cat /tmp/excluded-files.txt)
MAINT="./maintainers/flake-module.nix"

for file in $FILES
do
    shellcheck -f diff "$file" | git apply
    SC=$?

    # if shellcheck/git patch fails.... manually fix
    if [ $SC -ne 0 ]; then
        echo "$file failed to be formatted manual intervention required"
        nvim $file
    else
        # Remove the processed file from the list
        #
        # we need to escape the nested directories for our regex.
        escaped_file=$(echo "$file" | sed -e 's/[]\/$*.^[]/\\&/g')
        sed -i "/$escaped_file/d" $MAINT

        # Do git stuffs
        git add "$file" "$MAINT"
        git commit -m "housekeeping: shellcheck for $file"
    fi
done
```

**Comment by Ericson2314** [comment]
@SkamDart Yes there is! See https://nixos.org/manual/nix/unstable/contributing/testing and `./mk/debug-test.sh tests/functional/${testName}.sh`.

**Comment by bryanhonof** [comment]
What should we do about scripts like the following?

https://github.com/NixOS/nix/blob/838b666a809e38e741aa56e57a31dd035b55fefb/tests/functional/shell.shebang.sh#L1-L4

I can't disable [SC2230](https://www.shellcheck.net/wiki/SC2239), or at least, I don't think I can.

**Comment by Ericson2314** [comment]
@bryanhonof we might be able to put some disables in the precommit configuration as a workaround?

**Comment by Ericson2314** [comment]
Per the discussion in https://github.com/NixOS/nix/pull/10857, on issue with shellcheck's `--diff` is is https://github.com/koalaman/shellcheck/issues/2996, quotes appears in the middle of "words" rather than next to spaces, is a funky style.

In https://github.com/NixOS/nix/pull/10857#issuecomment-2163483137 I through together this hackjob sed (not fully correct, just a heuristic!) that can help a big fixing.
```sh
sed -i -E -e 's/( "[^"]+)"([a-z/=#_1-9-]+)( |\)|$)/\1\2"\3/' tests/functional/**.sh
```

**Comment by Ericson2314** [comment]
This is done now
